src/utils/api.ts

```diff
@@ -1,29 +1,67 @@
-import { Pokemon, PokemonData, PokemonDetails } from 'types';
+import { ErrorStatuses } from 'appConstants';
 
-export const fetchPokemons = async (url: string) => {
-  let pokemons: Array<Pokemon> = [];
+export const isServerError = (status: number) => String(status).startsWith(ErrorStatuses.server);
+
+export const hasError = (status: number) => !String(status).startsWith(ErrorStatuses.success);
+
+const handleCatch = (error: unknown) => {
+  error instanceof Error && window.alert(error.message);
+};
+
+export const handleResponse = async <T>(response: Response) => {
+  if (hasError(response.status)) {
+    if (isServerError(response.status)) {
+      window.alert('There was a server error');
+    }
+  } else return (await response.json()) as T;
+};
+
+export const handleMappedResponse = async <T>(arr: Array<Promise<Response>>) => {
+  const responses = await Promise.allSettled(arr);
+  const errors = responses.filter((res) => res.status === 'rejected').length;
+  const fullfilled = responses.filter((res) => res.status === 'fulfilled') as Array<
+    PromiseFulfilledResult<Response>
+  >;
+  const data = fullfilled.map((res) => res.value.json() as Promise<T>);
+  if (errors > 0) window.alert(`There was ${errors} errors on ${arr.length} requests.`);
+  return await Promise.all(data);
+};
+
+export const fetchPokemonByParameter = async <T>(url: string): Promise<T | null> => {
+  let results: T | null = null;
+  try {
+    const response = await fetch(url);
+    const data = await handleResponse<T>(response);
+    if (data) results = data;
+  } catch (error) {
+    handleCatch(error);
+  } finally {
+    return results;
+  }
+};
+
+export const fetchBase = async <T1 extends { results: Array<T2> }, T2>(url: string) => {
+  let base: Array<T2> = [];
   try {
-    const data = await fetch(url);
-    const { results }: PokemonData = await data.json();
-    pokemons = [...results];
+    const response = await fetch(url);
+    const data = await handleResponse<T1>(response);
+    if (data) base = [...data.results];
   } catch (error) {
-    console.error(error);
+    handleCatch(error);
   } finally {
-    return pokemons;
+    return base;
   }
 };
 
-export const fetchDetailedPokemons = async (pokemons: Array<Pokemon>) => {
-  let detailedPokemons: Array<PokemonDetails> = [];
+export const fetchDetails = async <T1 extends { url: string }, T2>(arr: Array<T1>) => {
+  let detailed: Array<T2> = [];
   try {
-    const responses = pokemons.map(async (pokemon) => {
-      const response = await fetch(pokemon.url);
-      return (await response.json()) as PokemonDetails;
-    });
-    detailedPokemons = await Promise.all(responses);
+    const responses = arr.map((item) => fetch(item.url).then((response) => response));
+    const data = await handleMappedResponse<T2>(responses);
+    detailed = data;
```

Comment:
why you can't return data from try block?

Reply:
Technically I could, of course. But if I know I'm fetching data that comes as an array then I'm used to returning empty array even if there was an error (handled). If I returned data in try block and left catch with handleCatch only, then returning type of my function would be Array<T2> | undefined. And component needs to check what comes.

It's not any pattern I saw in the web. It's something that I just started doing intuitively. If it's a bad way or there is a better one just let me know.

Reply:
It's uncommon practice. Your code should be predictable. Easy to read. If i see somrthing uncommon, i'm trying to find out reasons of this behaviour. IMHO finally block should works for like some sideeffects that should appear after request. Like notification that request finished. In try block you work with res. In catch block you handle an error. 
Simple. Predictable. 
If you want to get rid of undefined in result of function call, and in case of error, you want your app to get just empty array it is easy to do.
`const myArr = someFc('ololo') ?? [] as MyArrType;` (maybe works even without type assertion)
If you getting an object with array as a field - work become more easier then above.
`const {myArr=[]} = someFc('ololo');`

Reply:
Okay I changed it

Yeah, it works without type assertion. But the only inconvenience I notice here is putting the responsibility for checking data on the component. And each time you use the function you have to remember (or notice ts warn) to add checking code.